src/app/dashboard/dashboard.controller.js

```diff
@@ -18,61 +19,105 @@
 
-      if (!isUpdate) {
-        copy.title = getUniqueName(copy.title);
-      }
-
       return $mdDialog.show({
         controller: PluginConfigCtrl,
         controllerAs: 'vm',
         bindToController: true,
         templateUrl: 'app/dashboard/plugin-config.html',
-        locals: { plugin: copy }
+        locals: { plugin: copy, isUpdate: isUpdate }
       });
     }
 
     function PluginConfigCtrl() {
       var vm = this;
       var pluginTitle = vm.plugin.title;
+      var isUpdate = vm.isUpdate;
+
+      if (isUpdate) {
+        vm.save = save;
+      } else {
+        // Don't allow `save` until we have a unique name to suggest.
+        getUniquePlugin(vm).then(function() {
+          vm.save = save;
+        });
+
+      }
 
       vm.close = $mdDialog.hide;
-      vm.save = save;
 
       function save() {
-        return $dashboard
-          .updatePlugin(pluginTitle, vm.plugin)
-          .then(vm.close);
+        var id = vm.id || getSlug(pluginTitle);
+        var $plugin = $dashboard.getPlugin(id);
+
+        if (isUpdate) {
+          // We don't have to find a unique plugin ID.
+          return $plugin.save(vm.plugin).then(vm.close);
+        } else {
+          return getUniquePlugin(vm)
+            .then(function($plugin) {
+              return $plugin.save(vm.plugin);
+            })
+            .then(vm.close);
+        }
       }
     }
 
-    function clone(plugin) {
-      return configure(plugin, false);
-    }
+    function getUniquePlugin(vm) {
+      var id = vm.id || getSlug(vm.plugin.title);
+      var originalTitle = vm.plugin.title;
 
-    function getUniqueName(name) {
-      if (!$dashboard.getPluginByTitle(name)) {
-        return name;
-      }
+      return $dashboard.read().then(function(data) {
+        function exists(id) {
+          return data && data.plugins && data.plugins[id];
+        }
 
-      var i = 0;
-      var title;
+        function resolvePlugin(id) {
+          var $plugin = $dashboard.getPlugin(id);
+          return $q.resolve($plugin);
+        }
+
+        if (!exists(id)) {
+          // This name is fine.
+          return resolvePlugin(id);
+        } else {
+          var i = 0;
 
-      while (++i) {
-        title = name + ' ' + i;
+          while (++i) {
+            vm.plugin.title = originalTitle + ' (' + i + ')';
+            id = getSlug(vm.plugin.title);
 
-        if (!$dashboard.getPluginByTitle(title)) {
-          return title;
+            if (!exists(id)) {
+              return resolvePlugin(id);
+            }
+          }
         }
-      }
+      });
     }
 
-    function remove(plugin) {
-      return $dashboard
-        .removePlugin(plugin)
+    function clone(pluginId, plugin) {
+      return configure(plugin, false);
+    }
+
+    function remove(pluginId, plugin) {
+      var $plugin = $dashboard.getPlugin(pluginId);
+
+      return $plugin.delete()
+        .then(function() {
+          // Return to the dashboard view.
```

Comment:
Ok, so what if one day after I've cluttered my portal with over 9000 plugins, I decide to go on a cleaning spree. Since I have the ability to delete plugins that I'm not currently looking at (or did that change and I haven't realized it yet) does it make sense to redirect the user after deleting one?